lib/IDE/CompletionInstance.cpp

```diff
@@ -276,6 +276,36 @@ static bool areAnyDependentFilesInvalidated(
       });
 }
 
+/// Get interface hash of \p SF including the type members in the file.
+///
+/// This is needed because SouceFile::getIntefaceHash() doesn't digest type body
+/// fingerprints in the file.
+void getInterfaceHashIncludingTypeMembers(SourceFile *SF,
+                                          llvm::SmallString<32> &str) {
+  /// FIXME: Gross. Hashing multiple "hash" values.
+  llvm::MD5 hash;
+  SF->getInterfaceHash(str);
+  hash.update(str);
+
+  std::function<void(IterableDeclContext *)> hashTypeBodyFingerprints =
+      [&](IterableDeclContext *IDC) {
+        if (auto fp = IDC->getBodyFingerprint())
+          hash.update(*fp);
+        for (auto *member : IDC->getParsedMembers())
+          if (auto *childIDC = dyn_cast<IterableDeclContext>(member))
+            hashTypeBodyFingerprints(childIDC);
+      };
+
+  for (auto *D : SF->getTopLevelDecls()) {
+    if (auto IDC = dyn_cast<IterableDeclContext>(D))
+      hashTypeBodyFingerprints(IDC);
+  }
+
+  llvm::MD5::MD5Result result;
+  hash.final(result);
+  str = result.digest();
+}
+
```

Comment:
Couple of thoughts:
1. A bit more of the "why" in the D'Oxygen comment could help. Why does completion need this? So it can use a cache for the whole file, I think.
2. I'm not sure this will pick up every type-fingerprint. Will it work on arbitrary expressions that contain closures that contain a `struct`, for instance? What about an expression containing an interpolated string containing a closure containing a `struct`?

Reply:
1. Thanks for suggestion. I will update the comment.
2. In this use case, local types and their members aren't considered "interface". They are not visible from outside.

Reply:
Updated the doc-comment

Reply:
What if I’m doing a completion in a function in a struct in a closure? Is it still ok to miss a change in that struct?
I’ll change my review to not block you so you can merge if this case won’t be a problem.

Reply:
For fast-completion inside function bodies, the whole function body is re-parsed. So a struct in closure (which is in the function body) is "refreshed" anyway.

Reply:
Thanks for explaining. That would make a great comment in some future PR, if it's not there already.